### PR TITLE
Enable hardcoding gas in scripts

### PR DESCRIPTION
**Pull request by ckoopmann**
## Motivation

Currently when a user sets the gas limit of a broadcasted contract call with `someContract.someFunction{gas: 123456}()` the specified gas limit is used in the evm simulation but when the transaction is actually broadcasted on chain it is overwritten based on the estimated gas.

Therefore it is currently not possible to set a fixed gas limit for a single call in a script.

(closes: https://github.com/foundry-rs/foundry/issues/2627 )

## Solution

Persist the user specified gas limit (which is returned by `revm` in the `CallInput` struct), and exclude the given transaction from all "gas estimation" steps later on.

### CODE REVIEW

**Comment by ckoopmann** [comment]
The main challenge is how to determine if the `gas_limit` returned from the evm has been set by the user or is the result of estimation. 
I have not found a way to reliably extract this information anywhere, however I have observed that in the "estimation" case the gas limit seems to be set to the "remaining gas" which is the user specified total "transaction gas limit" minus any gas spent by previous transactions.

Therefore I have added a heuristic (see: https://github.com/foundry-rs/foundry/pull/4219/files#diff-4fec2ffcb2bf6a97025c3409d69b954d3a01035d7240e19e47f9e8ad95a5a32fR327)  that assumes that the call gas limit has been set manually if it is lower than the block gas limit AND the configured transaction gas limit is higher than the block limit.

While this should work reasonably well with the default configuration which sets the transaction limit to `i32.max`, it would result in unexpected behaviour if:

1. The user sets the transaction gas limit lower than the block gas limit. (In this case all manual gas limits are ignored).
2. Tx gas limit is set slightly higer than block gas limit but the remaining gas sinks below the block gas limit after a few of the transactions. (In this case estimated gas limits will be misinterpreted as fixed ones and wrongly excluded from later re-estimation).

An alternative heuristic I had in mind was adding a new configuration parameter `max_fixed_gas_limit` set by the user and using this limit instead of the `block_gas_limit` to determine which gas limits where manually set.

However the best option would be to actually determine this from some compilation artifact, source code, evm output etc.. 
Any Feedback or ideas on this would be much appreciated.

**Review by gakonst — CHANGES_REQUESTED** [review]

File: cli/src/cmd/forge/script/broadcast.rs [hunk 1]
```diff
@@ -237,9 +241,11 @@ impl ScriptArgs {
 
                 // Chains which use `eth_estimateGas` are being sent sequentially and require their
                 // gas to be re-estimated right before broadcasting.
-                if has_different_gas_calc(provider.get_chainid().await?.as_u64()) ||
-                    self.skip_simulation
+                if !is_fixed_gas_limit &&
+                    (has_different_gas_calc(provider.get_chainid().await?.as_u64()) ||
+                        self.skip_simulation)
                 {
+                    // TODO: Check that this case also works when setting gas limit for tx manually
```
[the remark this answers is not in the document]
Comment (ckoopmann) [reply cli/src/cmd/forge/script/broadcast.rs:248]: In the latest version I'm pretty sure that _it_ (bypassing gas estimation for fixed gas limits for chains with `has_different_gas_calc(..) = true`) works, but I will see if I can reproduce this case in the tests.

File: cli/src/cmd/forge/script/executor.rs [hunk 2]
```diff
@@ -182,8 +182,15 @@ impl ScriptArgs {
                         runner.executor.env_mut().block.number += U256::one();
                     }
 
-                    // We inflate the gas used by the user specified percentage
-                    tx.gas = Some(U256::from(result.gas_used * self.gas_estimate_multiplier / 100));
+                    let is_fixed_gas_limit = tx.gas.is_some();
```
[the remark this answers is not in the document]
Comment (ckoopmann) [reply cli/src/cmd/forge/script/executor.rs:185]: At this point `tx.gas` is what was set by the inspector code [here](https://github.com/ckoopmann/foundry/blob/12d698ebe28ca467a9a93c0cf4f48ecc04b56f74/evm/src/executor/inspector/cheatcodes/mod.rs#L450).  
So if the gas_limit was  set manually then it will be that value otherwise it will be `none` (which is always the case in the current implementation on master).

File: evm/src/executor/inspector/cheatcodes/util.rs [hunk 3]
```diff
@@ -324,6 +327,20 @@ pub fn parse_private_key(private_key: U256) -> Result<SigningKey, Bytes> {
...
+// not be overwritten by later estimations
+pub fn check_if_fixed_gas_limit<DB: DatabaseExt>(
+    data: &EVMData<'_, DB>,
+    call_gas_limit: u64,
+) -> bool {
+    // If the gas limit was not set in the source code it is set to the estimated gas left at the
+    // time of the call, which should be rather close to configured gas limit.
+    // TODO: Find a way to reliably make this determenation. (for example by
+    // generating it in the compilation or evm simulation process)
+    U256::from(data.env.tx.gas_limit) > data.env.block.gas_limit &&
```
[the remark this answers is not in the document]
Comment (ckoopmann) [reply evm/src/executor/inspector/cheatcodes/util.rs:340]: I don't think that would work due to the following reason:

The call_gas_limit (which comes from the CallInput struct returned by revm is always set to some value. In the case of "estimated" gas it will be set to all the remaining gas at the time of the call as determined by revm. From my understanding this is calculated based on the configured gas_imit (which is set to i32.max by default and does not take the senders eth balance into account).

So if we just always pass through the gas_limit we will end up with values that are way too high and corresponding transaction failures when submitting the transaction.

Basically this is the issue I tried to describe (way too verbosely) in my [comment here](https://github.com/foundry-rs/foundry/pull/4219#issuecomment-1407601873).

[the remark this answers is not in the document]
Comment (ckoopmann) [reply evm/src/executor/inspector/cheatcodes/util.rs:340]: In short the answer is: "Yes we need this function, but we should probably find a better way to determine wether the gas_limit was set manually."

File: evm/src/executor/inspector/cheatcodes/mod.rs [hunk 4]
```diff
@@ -445,6 +447,11 @@ where
                                 value: Some(call.transfer.value),
                                 data: Some(call.input.clone().into()),
                                 nonce: Some(account.info.nonce.into()),
+                                gas: if is_fixed_gas_limit {
+                                    Some(call.gas_limit.into())
+                                } else {
+                                    None
+                                },
```
[the remark this answers is not in the document]
Comment (ckoopmann) [reply evm/src/executor/inspector/cheatcodes/mod.rs:454]: Note that we are setting the `gas` field on the `TransactionRequest` struct at the time of its construction here.
So we are not "overriding" anything on the call object. 

I can see if there is a way to override the field on the `call` object instead (as soon as possible to its construction), but I feel it might have a higher chance of unintended side effects elsewhere.

If this is purely about code aesthetics I could also replace the `check_if_fixed_gas_limit` helper function with `get_gas_limit_if_fixed` and change this here to:
`gas: get_gas_limit_if_fixed(data, call.gas_limit)`.

File: cli/src/cmd/forge/script/broadcast.rs (hunk 1 above)
Comment (mattsse) [inline cli/src/cmd/forge/script/broadcast.rs:248]: is this still a todo or covered via test?

**Review by mattsse — COMMENTED** [review]
sorry for the wait @ckoopmann 
somehow this one slipped through...

imo this is reasonable change.

lgtm, pending q regarding the TODO

File: cli/src/cmd/forge/script/broadcast.rs (hunk 1 above)
Comment (ckoopmann) [reply cli/src/cmd/forge/script/broadcast.rs]: I added a test that should cover this and accordingly removed the TODO comment:
https://github.com/foundry-rs/foundry/pull/4219/commits/06b0bd04ad992ba62423b9c344854c2f31991b11

**Comment by ckoopmann** [comment]
> sorry for the wait @ckoopmann somehow this one slipped through...
> 
> imo this is reasonable change.
> 
> lgtm, pending q regarding the TODO

No problem, probably was my fault anyway since I apparently accidentally unassigned your review request. 
I addressed your question. Should be mergeable by now. 

Please note though that the logic detecting wether a gas limit coming from `revm` was a fixed limit set by the user or the result of internal estimation by `revm` is still a tiny bit hacky.
(See comments and discussions above).
Should work fine in 99% of cases but might be worth to open an issue to improve this in the future.

**Review by gakonst — APPROVED** [review]
good with me. thank you
